File: linked-lists/linked-list-oop.js
class Node {
  constructor(data, next = null) {
    this.data = data;
    this.next = next;
  }

  prepend(value) {
    return new Node(value, this);
  }

  unprepend() {
    return this.next;
  }
}

class LinkedList {
  constructor() {
      this.head = null;
      this.tail = null;
      this.size = 0;
  }

  addToFront(value) {
      this.head = new Node(value, this.head);

      if (this.size === 0) {
        this.tail = this.head;
      }

      this.size++;

      return this;
  }

  removeFromFront() {
    if (!this.head) {
      return null;
    }

    let value = this.head.value;
    this.head = this.head.unprepend();

    this.size--;

    if (this.size === 0) {
      this.tail = null;
    }

    return value;
  }

  addToTail(value) {
    let newTail = new Node(value);

    if (this.size === 0) {
      this.head = newTail;
    } else {
      this.tail.next = newTail;
    }

    this.size++;

    this.tail = newTail;
  }
}

module.exports = {Node, LinkedList};

File: linked-lists/linked-list-oop.py
class Node:
    def __init__(self, data, next=None):
        self.data = data
        self.next = next

    def prepend(self, value):
        return Node(value, self)

    def unprepend(self):
        return self.next

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def add_to_front(self, value):
        new_node = Node(value, self.head)
        self.head = new_node

        if self.size == 0:
            self.tail = self.head

        self.size += 1

        return self

    def remove_from_front(self):
        if not self.head:
            return None

        value = self.head.data
        self.head = self.head.unprepend()

        self.size -= 1

        if self.size == 0:
            self.tail = None

        return value

    def add_to_tail(self, value):
        new_tail = Node(value)

        if self.size == 0:
            self.head = new_tail
        else:
            self.tail.next = new_tail

        self.size += 1

        self.tail = new_tail
